python/semantic_kernel/kernel.py

```diff
@@ -408,51 +297,121 @@ async def invoke_prompt(
     # endregion
     # region Function Invoking/Invoked Events
 
-    def on_function_invoked(
+    async def pre_function_invoke(
```

Comment:
Should we keep it as private?

Reply:
in python nothing is actually private, and this is called as `kernel.pre_function_invoke` from KernelFunction, so typing would complain about calling a private method.

Reply:
`in python nothing is actually private` - I meant calling it like `__pre_function_invoke`. Just want to make sure that users won't call this method manually. If it's not possible to rename it, maybe we need to consider putting this method somewhere outside of kernel, so it's not accessible.

Reply:
yeah, so calling it `_pre_function_invoke` will make the type checkers complain in KernelFunction when it is calling a 'private' method on Kernel, and I don't really see the harm in people calling this manually...

Reply:
> I don't really see the harm in people calling this manually

It depends how you design public API in Python. Calling it manually shouldn't provide any benefits, so it could be confusing for users what is the purpose of this method, especially on kernel level. In .NET version we expose only required methods and hide everything else. It also helps to automatically generate a documentation based on public API and avoid documenting internal/private methods. 

If type checkers will complain in KernelFunction, you can move these methods from Kernel to KernelFunction. It still should work because in KernelFunction you should have access to kernel. 

I will leave it up to you, but it's better to hide methods that shouldn't be called manually.